File: backend/app/schemas/interview.py
"""Request/response shapes for the mock-interview module (Phase 8).
`average_score` is never stored — routes compute it from `answers` at
response time, same reasoning as the career module's on-the-fly scoring:
there's nothing to keep in sync with a cache.
"""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.interview_question import QuestionCategory, QuestionDifficulty
from app.models.mock_interview import SessionStatus
from app.schemas.career import CareerRoleSummary
from app.schemas.profile import SkillPublic


class InterviewQuestionPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    question_text: str
    category: QuestionCategory
    difficulty: QuestionDifficulty
    skills: list[SkillPublic]

    @field_validator("skills", mode="before")
    @classmethod
    def _unwrap_question_skills(cls, value):
        # InterviewQuestion.skills is a list of InterviewQuestionSkill join
        # rows; unwrap to the underlying Skill, same pattern as
        # ProjectPublic._unwrap_project_skills in schemas/profile.py.
        return [item.skill if hasattr(item, "skill") else item for item in value]


class MockInterviewAnswerPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    question_id: uuid.UUID
    answer_text: str
    ai_feedback: list[str] | None
    ai_score: int | None
    ai_provider_used: str | None
    created_at: datetime


class MockInterviewAnswerResponse(BaseModel):
    data: MockInterviewAnswerPublic


class StartSessionRequest(BaseModel):
    career_role_id: uuid.UUID | None = None
    question_count: int = Field(default=5, ge=3, le=10)


class SubmitAnswerRequest(BaseModel):
    question_id: uuid.UUID
    answer_text: str = Field(min_length=10, max_length=4000)


class MockInterviewSessionSummary(BaseModel):
    id: uuid.UUID
    career_role: CareerRoleSummary | None
    status: SessionStatus
    question_count: int
    answered_count: int
    average_score: int | None
    created_at: datetime
    completed_at: datetime | None


class MockInterviewSessionListResponse(BaseModel):
    data: list[MockInterviewSessionSummary]


class MockInterviewSessionDetail(BaseModel):
    id: uuid.UUID
    career_role: CareerRoleSummary | None
    status: SessionStatus
    questions: list[InterviewQuestionPublic]
    answers: list[MockInterviewAnswerPublic]
    average_score: int | None
    created_at: datetime
    completed_at: datetime | None


class MockInterviewSessionResponse(BaseModel):
    data: MockInterviewSessionDetail

File: frontend/types/interview.ts
import type { CareerRoleSummary } from "./career";
import type { Skill } from "./profile";

export type QuestionCategory = "behavioral" | "technical" | "situational";
export type QuestionDifficulty = "easy" | "medium" | "hard";
export type SessionStatus = "in_progress" | "completed";

export interface InterviewQuestion {
  id: string;
  question_text: string;
  category: QuestionCategory;
  difficulty: QuestionDifficulty;
  skills: Skill[];
}

export interface MockInterviewAnswer {
  id: string;
  question_id: string;
  answer_text: string;
  ai_feedback: string[] | null;
  ai_score: number | null;
  ai_provider_used: string | null;
  created_at: string;
}

export interface MockInterviewSessionSummary {
  id: string;
  career_role: CareerRoleSummary | null;
  status: SessionStatus;
  question_count: number;
  answered_count: number;
  average_score: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface MockInterviewSessionDetail {
  id: string;
  career_role: CareerRoleSummary | null;
  status: SessionStatus;
  questions: InterviewQuestion[];
  answers: MockInterviewAnswer[];
  average_score: number | null;
  created_at: string;
  completed_at: string | null;
}
